### PR TITLE
Upgrade MDX 

### PR DESCRIPTION
**Pull request by jeffsee55**
Fixes https://github.com/tinacms/tinacms/issues/2420
Fixes https://github.com/tinacms/tinacms/issues/2469
Fixes https://github.com/tinacms/tinacms/issues/2315
Fixes https://github.com/tinacms/tinacms/issues/2284
Fixes https://github.com/tinacms/tinacms/issues/2270
Fixes https://github.com/tinacms/tinacms/issues/2131

---

- [x] Publish tag (`yarn add tinacms@upgrade-to-plate-9`)
- [x] Update [a branch](https://github.com/tinacms/tina-cloud-starter/pull/218) in tina-cloud-starter to use the feature
- [x] Update [a branch](https://github.com/tinacms/tina-barebones-starter/pull/23) in basic starter to use the feature
- [x] Update [a branch](https://github.com/tinacms/tina-docs-starter/pull/33) in docs starter to use the feature

### CODE REVIEW

**Comment by logan-anderson** [comment]
I know this is still a WIP just so you are aware I am getting this error when adding a code block to exciting tex
Screen Shot 2022-01-25 at 1 46 51 PM
t

**Comment by spbyrne** [comment]
Screen Shot 2022-01-25 at 3 01 13 PM

This is what you should be seeing.

**Comment by logan-anderson** [comment]
Another piece of feedback. 

When there is no templates the dropdown is blank. Maybe we could have a some text saying "no components. Click here to learn how to add them" or something along those lines.
Screen Shot 2022-01-25 at 1 59 15 PM

**Comment by spbyrne** [comment]
@logan-anderson I pushed a fix today that hides the embed button when there's no templates.

**Comment by logan-anderson** [comment]
> That looks like the old UI @logan-anderson, are you sure that's a recent build?

I did not have the new editor enabled. it is might be a parsing issue with both of the editors (new and old) does not have to be addressed in this PR I just noticed it.

File: packages/@tinacms/toolkit/src/packages/fields/plugins/MdxFieldPlugin/updated/plate/plugins/create-slash-plugin/index.tsx [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+    Transforms.removeNodes(props.editor, {
+      match: (n) => {
+        // @ts-ignore bad type from slate
+        if (Element.isElement(n) && n.type === ELEMENT_MAYBE_MDX) {
+          return true
+        }
+      },
+    })
+    insertMDX(props.editor, value)
+  }
```
Comment (jeffsee55) [inline packages/@tinacms/toolkit/src/packages/fields/plugins/MdxFieldPlugin/updated/plate/plugins/create-slash-plugin/index.tsx:113]: Once we've selected a JSX embed, [unwrap](https://docs.slatejs.org/api/transforms#transforms.unwrapnodes-editor-editor-options) the `ELEMENT_MAYBE_MDX` node. When we first type `/`, the element is turned into an `ELEMENT_MAYBE_MDX` node (that happens [here](https://github.com/tinacms/tinacms/pull/2473/files#diff-e72a6523ea2b2a181086ea7b625bfdc84d0bdc2e69c70ebbe48e9a23bee101d2R64-R66)):
Screen Shot 2022-01-27 at 3 18 34 PM

```
[
  {
    "type": "p",
    "children": [
      {
        "text": "Hello, "
      },
      {
        "type": "maybe_mdx",
        "children": [
          {
            "type": "text",
            "text": "/Da",
          }
        ],
      },
      {
        "text": ""
      }
    ],
  }
]
```
And then the `insertMDX` action kicks in, inserting the 'Date & Time' element:

Screen Shot 2022-01-27 at 3 19 29 PM

```
[
  {
    "type": "p",
    "children": [
      {
        "text": "Hello, "
      },
      {
        "type": "mdxJsxTextElement",
        "name": "DateTime",
        "children": [
          {
            "text": ""
          }
        ],
        "props": {},
      },
      {
        "text": ""
      }
    ],
  }
]
```

File: packages/@tinacms/toolkit/src/packages/fields/plugins/MdxFieldPlugin/updated/plate/plugins/ui/combobox.tsx [hunk 2]
```diff
@@ -0,0 +1,225 @@
...
+/**
+ * This could (should?) be replaced by something like downshift, but in some
+ * ways it's not like a traditional combobox because so much of what's needed
+ * here should _not_ steal focus from the editor. Some of the combobox libraries
+ * I've tried really shine when it comes to accessibility out-of-the-box which is
+ * sort of the opposite of what we want here. I'm not sure of the best way to make
+ * this accessible, but I have a feeling the plate combobox is the way to go.
+ * At first glance, there's very little documentation for using it, though it
+ * does seem to support bringing out own UI (it uses downshift).
+ */
```
Comment (jeffsee55) [inline packages/@tinacms/toolkit/src/packages/fields/plugins/MdxFieldPlugin/updated/plate/plugins/ui/combobox.tsx:127]: See comment

**Review by jeffsee55 — COMMENTED** [review]
Known limitations

- Drag-n-drop for images doesn't work
- Code block deletion is buggy
  - Not possible when it's the first item
  - When it's the last item, code_line lingers
- Block quote doesn't accept other block elements (JSX, headers, etc)

**Review by logan-anderson — APPROVED** [review]
In terms of code quality and usability this looks good to me.

There is a couple of small UI stuff that is not super important and does not have to block this from being merged. should I file those as GH issues with a MDX tag?

**Comment by jeffsee55** [comment]
> There is a couple of small UI stuff that is not super important and does not have to block this from being merged. should I file those as GH issues with a MDX tag?

You can mention it here and if it's low-hanging fruit I can knock it out

**Comment by logan-anderson** [comment]
Heading of level 4,5,6 and paragraph all of the same styling so its very hard to tell them apart in the editor.

Screen Shot 2022-01-31 at 4 11 48 PM

**Comment by logan-anderson** [comment]
Right away (without making any edits) it shows the "Unsaved Changes" indicator.

Screen Shot 2022-01-31 at 4 31 47 PM

**Comment by jeffsee55** [comment]
> Heading of level 4,5,6 and paragraph all of the same styling so its very hard to tell them apart in the editor.

I've set h5 to `13px` and h6 to `12px` (cc @spbyrne)

> Right away (without making any edits) it shows the "Unsaved Changes" indicator.

I've improved this, but not fixed it entirely. Now you should only see that indicator change once you interact with the rich-text area. Note that it will trigger the indicator just by the act of clicking on the area. The reason for this is that we need to normalize so they always have `children`, even if they're void nodes. The drawback of this is that the shape of the editor changes as soon as we interact with it (because the normalizer runs). This is something I'll have to think more about so I've created a ticket.

**Comment by logan-anderson** [comment]
Noticed a small UI issue where I could not select from the dropdown because the code block interfered. 
Screen Shot 2022-02-01 at 9 30 19 AM

**Comment by spbyrne** [comment]
Fixed header sizes in the editor & ddl. H1 is quite large but also should pretty much never be used as a content heading.
Screen Shot 2022-02-01 at 1 42 47 PM
Screen Shot 2022-02-01 at 1 42 59 PM

File: packages/@tinacms/toolkit/src/packages/fields/plugins/MdxFieldPlugin/index.tsx [hunk 3]
```diff
@@ -16,9 +16,18 @@ limitations under the License.
...
 import { RichEditor } from './plate'
+import { RichEditor as RichEditorUpdated } from './updated'
+import { useCMS } from '../../../react-core'
 
 export const MdxFieldPlugin = {
   name: 'rich-text',
-  Component: RichEditor,
+  Component: (props) => {
+    const cms = useCMS()
+    if (cms.flags.get('rich-text-alt')) {
```
[the remark this answers is not in the document]
Comment (Enigmatical) [reply packages/@tinacms/toolkit/src/packages/fields/plugins/MdxFieldPlugin/index.tsx:28]: Maybe `improved-rich-text` or `upgrade-rich-text`, but it is a temporary flag, so not a big deal.

**Comment by jeffsee55** [comment]
Fixing a few things from @spbyrne from https://github.com/tinacms/product-roadmap/issues/13#issuecomment-1028059635
- [x] Clicking combobox slate point error
- [x] Block embed swallowing current block
- [x] Save sometimes removing all content
- [x] Missing title on links

Some of the link selection stuff is tricky, so this is what I'm working on now https://github.com/tinacms/product-roadmap/issues/13#issuecomment-1028126204
- [ ] Link form when no selection
- [ ] Strange link selection behavior

**Review by logan-anderson — APPROVED** [review]
🎉

**Pull request merged**